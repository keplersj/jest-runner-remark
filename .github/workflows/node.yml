name: Node CI

on: [push]

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4
      - uses: actions/setup-node@v4
        with:
          node-version: lts/*
      - name: Install npm dependencies
        run: npm ci
      - name: Run tests
        run: npm test -- --ci
      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v4.6.0
