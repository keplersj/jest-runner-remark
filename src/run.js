const engine = require("unified-engine");
const remark = require("remark");
const stream = require("stream");
const { pass, fail } = require("create-jest-runner");

class NoOpStream extends stream.Writable {
  _write(chunk, enc, next) {
    next();
  }
}

module.exports = ({ testPath }) => {
  const output = [];

  class StoreStream extends stream.Writable {
    _write(chunk, enc, next) {
      output.push(chunk.toString());
      next();
    }
  }

  const start = new Date();
  return new Promise((resolve, reject) => {
    engine(
      {
        processor: remark,
        files: [testPath],
        extensions: ["md", "mdx", "markdown", "mkd", "mkdn", "mkdown"],
        pluginPrefix: "remark",
        rcName: ".remarkrc",
        packageField: "remarkConfig",
        ignoreName: ".remarkignore",
        color: true,
        streamOut: new NoOpStream(),
        streamError: new StoreStream()
      },
      (error, code, context) => {
        if (error) {
          return reject(error);
        }

        if (code === 0) {
          const result = pass({
            start,
            end: new Date(),
            test: {
              path: testPath,
              title: "Remark",
              errorMessage: output.toString()
            }
          });

          if (context.files[0].messages.length != 0) {
            result.console = [
              { message: output.join(), origin: "", type: "warn" }
            ];
          }

          return resolve(result);
        } else if (code === 1) {
          return resolve(
            fail({
              start,
              end: new Date(),
              test: {
                path: testPath,
                title: "Remark",
                errorMessage: output.join()
              }
            })
          );
        } else {
          return reject();
        }
      }
    );
  });
};
