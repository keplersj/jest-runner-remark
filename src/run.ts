import { TestResult } from "@jest/test-result";
import { engine } from "unified-engine";
import { remark } from "remark";
import { Writable } from "stream";
import { pass, fail } from "create-jest-runner";

class NoOpStream extends Writable {
  _write(_: any, __: any, next: Function): void {
    next();
  }
}

interface Parameters {
  testPath: string;
}

export default ({ testPath }: Parameters): Promise<TestResult> => {
  const output: string[] = [];

  class StoreStream extends Writable {
    _write(chunk: any, _: any, next: Function): void {
      output.push(chunk.toString());
      next();
    }
  }

  const start = Date.now();
  return new Promise((resolve, reject) => {
    engine(
      {
        processor: remark,
        files: [testPath],
        extensions: ["md", "mdx", "markdown", "mkd", "mkdn", "mkdown"],
        pluginPrefix: "remark",
        rcName: ".remarkrc",
        packageField: "remarkConfig",
        ignoreName: ".remarkignore",
        color: true,
        streamOut: new NoOpStream(),
        streamError: new StoreStream(),
      },
      (error, code, context) => {
        if (error) {
          return reject(error);
        }

        if (code === 0) {
          const result = pass({
            start,
            end: Date.now(),
            test: {
              path: testPath,
              title: "Remark",
              errorMessage: output.toString(),
            },
          });

          if (
            context &&
            context.files &&
            context.files[0].messages.length != 0
          ) {
            result.console = [
              { message: output.join(), origin: "", type: "warn" },
            ];
          }

          return resolve(result);
        } else if (code === 1) {
          return resolve(
            fail({
              start,
              end: Date.now(),
              test: {
                path: testPath,
                title: "Remark",
                errorMessage: output.join(),
              },
            })
          );
        } else {
          return reject();
        }
      }
    );
  });
};
