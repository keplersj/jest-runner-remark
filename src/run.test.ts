import run from "./run";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

// Remove undeterministic data from test reports
expect.addSnapshotSerializer({
  print: (value: any, serialize) => {
    delete value.perfStats;
    delete value.testFilePath;
    for (const result of value.testResults) {
      delete result.duration;
    }
    return serialize(value);
  },
  test: (value) =>
    value && value.perfStats && value.testFilePath && value.testResults,
});

describe("Remark-Lint Jest Runner", () => {
  describe("good fixture", () => {
    it("matches snapshot", async () => {
      const result = await run({
        testPath: join(__dirname, "__fixtures__", `good.md`),
      });
      return expect(result).toMatchSnapshot();
    });
  });
});
