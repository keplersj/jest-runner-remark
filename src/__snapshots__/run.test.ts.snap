// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remark-Lint Jest Runner good fixture matches snapshot 1`] = `
Object {
  "console": undefined,
  "coverage": undefined,
  "displayName": undefined,
  "failureMessage": undefined,
  "leaks": false,
  "memoryUsage": undefined,
  "numFailingTests": 0,
  "numPassingTests": 1,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "openHandles": Array [],
  "skipped": false,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "uncheckedKeys": Array [],
    "unmatched": 0,
    "updated": 0,
  },
  "testExecError": undefined,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureDetails": Array [],
      "failureMessages": Array [
        "[4m[32msrc/__fixtures__/good.md[39m[24m: no issues found
",
      ],
      "fullName": "",
      "numPassingAsserts": 1,
      "status": "passed",
      "title": "Remark",
    },
  ],
  "v8Coverage": undefined,
}
`;
